src/backward.rs

```diff
@@ -34,41 +39,70 @@ impl From<u64> for NodeType {
+    }
     match node_type {
-        NodeType::Null => Value::Null,
+        NodeType::Null => Ok(Value::Null),
         NodeType::Boolean => {
             let buffer = raw::load_string_buffer(rdb);
-            Value::Bool(buffer.as_ref()[0] == b'1')
+            if is_io_error(rdb) {
+                return Err(Error::from("short read"));
+            }
+            Ok(Value::Bool(buffer.as_ref()[0] == b'1'))
+        }
+        NodeType::Integer => {
+            let n = raw::load_signed(rdb);
+            if is_io_error(rdb) {
+                return Err(Error::from("short read"));
+            }
+            Ok(Value::Number(n.into()))
+        }
+        NodeType::Number => {
+            let n = raw::load_double(rdb);
+            if is_io_error(rdb) {
+                return Err(Error::from("short read"));
+            }
+            Ok(Value::Number(Number::from_f64(n).unwrap()))
         }
-        NodeType::Integer => Value::Number(raw::load_signed(rdb).into()),
-        NodeType::Number => Value::Number(Number::from_f64(raw::load_double(rdb)).unwrap()),
         NodeType::String => {
             let buffer = raw::load_string_buffer(rdb);
-            Value::String(buffer.to_string().unwrap())
+            if is_io_error(rdb) {
+                return Err(Error::from("short read"));
+            }
+            Ok(Value::String(buffer.to_string().unwrap()))
         }
         NodeType::Dict => {
             let len = raw::load_unsigned(rdb);
+            if is_io_error(rdb) {
+                return Err(Error::from("short read"));
+            }
             let mut m = Map::with_capacity(len as usize);
             for _ in 0..len {
-                let t: NodeType = raw::load_unsigned(rdb).into();
+                let u = raw::load_unsigned(rdb);
+                if is_io_error(rdb) {
+                    return Err(Error::from("short read"));
+                }
+                let t: NodeType = u.into();
                 if t != NodeType::KeyVal {
-                    panic!("Can't load old RedisJSON RDB");
+                    return Err(Error::from("Can't load old RedisJSON RDB"));
                 }
                 let buffer = raw::load_string_buffer(rdb);
-                m.insert(buffer.to_string().unwrap(), json_rdb_load(rdb));
+                m.insert(buffer.to_string().unwrap(), json_rdb_load(rdb)?);
             }
-            Value::Object(m)
+            Ok(Value::Object(m))
         }
         NodeType::Array => {
             let len = raw::load_unsigned(rdb);
             let mut v = Vec::with_capacity(len as usize);
             for _ in 0..len {
-                v.push(json_rdb_load(rdb))
+                let nested = json_rdb_load(rdb)?;
+                v.push(nested);
             }
-            Value::Array(v)
+            Ok(Value::Array(v))
         }
-        NodeType::KeyVal => panic!("Can't load old RedisJSON RDB"),
+        NodeType::KeyVal => Err(Error::from("Can't load old RedisJSON RDB")),
```

Comment:
how will you be able to recover from it and not panic?

Reply:
when load fails, replica reports to master which should resend rdb